src/SprykerEco/Zed/AvalaraTax/Business/Calculator/AbstractCartItemAvalaraTaxCalculator.php

```diff
@@ -0,0 +1,164 @@
+{
+    protected const KEY_TAX_RATE = 'rate';
+
+    /**
+     * @var \SprykerEco\Zed\AvalaraTax\Business\Executor\AvalaraTransactionExecutorInterface
+     */
+    protected $avalaraTransactionExecutor;
+
+    /**
+     * @var \SprykerEco\Zed\AvalaraTax\Dependency\Facade\AvalaraTaxToMoneyFacadeInterface
+     */
+    protected $moneyFacade;
+
+    /**
+     * @var \SprykerEco\Zed\AvalaraTax\Dependency\Service\AvalaraTaxToUtilEncodingServiceInterface
+     */
+    protected $utilEncodingService;
+
+    /**
+     * @param \SprykerEco\Zed\AvalaraTax\Business\Executor\AvalaraTransactionExecutorInterface $avalaraTransactionExecutor
+     * @param \SprykerEco\Zed\AvalaraTax\Dependency\Facade\AvalaraTaxToMoneyFacadeInterface $moneyFacade
+     * @param \SprykerEco\Zed\AvalaraTax\Dependency\Service\AvalaraTaxToUtilEncodingServiceInterface $utilEncodingService
+     */
+    public function __construct(
+        AvalaraTransactionExecutorInterface $avalaraTransactionExecutor,
+        AvalaraTaxToMoneyFacadeInterface $moneyFacade,
+        AvalaraTaxToUtilEncodingServiceInterface $utilEncodingService
+    ) {
+        $this->avalaraTransactionExecutor = $avalaraTransactionExecutor;
+        $this->moneyFacade = $moneyFacade;
+        $this->utilEncodingService = $utilEncodingService;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\CalculableObjectTransfer $calculableObjectTransfer
+     *
+     * @return void
+     */
+    public function calculateTax(CalculableObjectTransfer $calculableObjectTransfer): void
+    {
+        if (!$this->hasShipmentAddress($calculableObjectTransfer)) {
+            return;
+        }
+
+        $avalaraCreateTransactionResponseTransfer = $this->avalaraTransactionExecutor->executeAvalaraSalesOrderTransaction(
+            $calculableObjectTransfer
+        );
+
+        $calculableObjectTransfer->getOriginalQuoteOrFail()->setAvalaraCreateTransactionResponse($avalaraCreateTransactionResponseTransfer);
+        if (!$avalaraCreateTransactionResponseTransfer->getIsSuccessful()) {
+            return;
+        }
+
+        $this->calculateTaxForItemTransfers($calculableObjectTransfer->getItems(), $avalaraCreateTransactionResponseTransfer);
+    }
+
+    /**
+     * @param \ArrayObject|\Generated\Shared\Transfer\ItemTransfer[] $itemTransfers
+     * @param \Generated\Shared\Transfer\AvalaraCreateTransactionResponseTransfer $avalaraCreateTransactionResponseTransfer
+     *
+     * @return void
+     */
+    abstract protected function calculateTaxForItemTransfers(
+        ArrayObject $itemTransfers,
+        AvalaraCreateTransactionResponseTransfer $avalaraCreateTransactionResponseTransfer
+    ): void;
+
+    /**
+     * @param \Generated\Shared\Transfer\CalculableObjectTransfer $calculableObjectTransfer
+     *
+     * @return bool
+     */
+    protected function hasShipmentAddress(CalculableObjectTransfer $calculableObjectTransfer): bool
```

Comment:
* This method is a good candidate to be an abstract method. 
* Each class(single/multi shipment handler) will define it based on its own needs.

Reply:
Resolved.